# VALIDACIONES CON TRY/CATCH

Por el momento solo lo realizaremos con los int y los double. 

Nota: recordar que si queremos validar cadenas tendriamos que crear una funcion para validar los caracteres que nosotros le permitiremos usar al usuario.


## TRY CATCH PARA ENTEROS

Esta es una manera de hacerlo con try catch y sin necesidad de utilizar una funcion que tengamos que validar caracteres


In [1]:
import java.util.InputMismatchException;

Scanner leer = new Scanner(System.in);
int numero;
do {
    try {
        System.out.print("ingrese un numero de tipo INT: ");
        numero = leer.nextInt();
        break;
    } catch (InputMismatchException ex) {
        System.out.println("HA INGRESADO UN DATO INCORRECTO");
        leer.nextLine();
    }
} while (true);

ingrese un numero de tipo INT: hola
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo INT: como
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo INT: estas
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo INT: 123.3
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo INT: 1209


In [4]:
numero

1209

## TRY CATCH PARA DOUBLE.

Vemos que es lo mismo que los enteros nada mas que ahora nos permite ingresar numeros con coma

Nota: Recordar que matematicamente al poder ingresar numeros reales tambien podemos ingresar los tipos de numeros que tambien pertenecen a los reales. Los numeros enteros pertenecen a los reales y por lo tanto este try/catch sirve para ttodos los numeros menos los imaginarios

In [2]:
double numero2;
do {
    try {
        System.out.print("ingrese un numero de tipo DOUBLE o INT: ");
        numero2 = leer.nextDouble();
        break;
    } catch (InputMismatchException ex) {
        System.out.println("HA INGRESADO UN DATO INCORRECTO");
        leer.nextLine();
    }
} while (true);

ingrese un numero de tipo DOUBLE o INT: hoa
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo DOUBLE o INT: como
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo DOUBLE o INT: estas
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo DOUBLE o INT: todo bien?
HA INGRESADO UN DATO INCORRECTO
ingrese un numero de tipo DOUBLE o INT: 123 12


In [3]:
numero2

123.0

### POR SI PREGUNTAN QUE PASO?
Aclararemos que el metodo nextDouble() del objeto Scanner leer va a tomar los valores de la linea hasta que encuentre un espacio, o un enter... por eso es que mi variable numero2 es igual a 123 y no te tiro error, ya que el 12 lo ignora por completo al estar separado por un espacio.